File: java-codingtest/src/main/java/programers/level2/TargetNumber.java
package programers.level2;

public class TargetNumber {

    private static int result = 0;

    public static void main(String[] args) {

    }

    public int solution(int[] numbers, int target) {
        DFS(numbers, 0, 0, target);

        return result;
    }

    public void DFS(int[] numbers, int sum, int i, int target) {
        if (numbers.length == i) {
            if (sum == target) {
                result++;
            }
            return;
        }

        DFS(numbers, sum + numbers[i], i + 1, target);
        DFS(numbers, sum - numbers[i], i + 1, target);
    }

}

File: kotlin-codingtest/src/main/kotlin/com/soten/programers/level2/TargetNumber.kt
package com.soten.programers.level2

private var answer = 0

private fun solution(numbers: IntArray, target: Int): Int {
	targeting(target, 0, numbers, 0)
	return answer
}

fun targeting(target: Int, index: Int, numbers: IntArray, cycle: Int) {
	if (cycle == target && index == numbers.size) {
		answer++
	}

	if (index >= numbers.size) {
		return
	}

	val plusCycle = cycle + numbers[index]
	val minusCycle = cycle - numbers[index]

	targeting(target, index + 1, numbers, plusCycle)
	targeting(target, index + 1, numbers, minusCycle)
}

fun main() {
	println(solution(intArrayOf(1, 1, 1, 1, 1), 3))
}
